app/src/main/java/com/sunasterisk/musixmatch/data/source/local/PlaylistLocalDataSource.java

```diff
@@ -0,0 +1,205 @@
+    private static final int FIRST_ELEMENT_INDEX = 0;
+    private static final int ERROR_INDEX = -1;
+    private static PlaylistLocalDataSource sInstance;
+    private ContentResolver mContentResolver;
+
+    private PlaylistLocalDataSource(Context context) {
+        mContentResolver = context.getContentResolver();
+    }
+
+    public static PlaylistLocalDataSource getsInstance(Context context) {
+        if (sInstance == null) {
+            sInstance = new PlaylistLocalDataSource(context);
+        }
+        return sInstance;
+    }
+
+    @Override
+    public void getPlaylist(Callback<List<Playlist>> callback) {
+        AsyncTask<Void, Void, List<Playlist>> asyncTask = new AsyncTask<Void, Void, List<Playlist>>() {
+            private List<Playlist> mPlaylists;
+            private Exception mException;
+
+            @Override
+            protected List<Playlist> doInBackground(Void... voids) {
+                try {
+                    mPlaylists = getLocalPlaylist();
+                } catch (Exception e) {
+                    mException = e;
+                }
+
+                return mPlaylists;
+            }
+
+            @Override
+            protected void onPostExecute(List<Playlist> playlists) {
+                super.onPostExecute(playlists);
+                if (playlists != null) {
+                    callback.getDataSuccess(playlists);
+                } else {
+                    callback.getDataFailure(mException);
+                }
+
+            }
+        };
+        asyncTask.execute();
+    }
+
+    @Override
+    public Uri addSongToPlaylist(long id, Track track) {
+        if (id < 0) {
+            return null;
+        }
+        // Find the greatest PLAY_ORDER in the playlist
+        int lastPlaylistId = getLargestPos(id);
+        ContentValues contentValues = new ContentValues();
+        contentValues.put(MediaStore.Audio.Playlists.Members.PLAY_ORDER, ++lastPlaylistId);
+        contentValues.put(MediaStore.Audio.Playlists.Members.AUDIO_ID, track.getTrackId());
+        Uri uri = mContentResolver.insert(MediaStore.Audio.Playlists.Members.getContentUri(CONTENT_URI_NAME, id), contentValues);
+        mContentResolver.notifyChange(Uri.parse(CONTENT_URI_MEDIA), null);
+        return uri;
+    }
+
+    public int getLargestPos(long id) {
+        Uri uri = MediaStore.Audio.Playlists.Members.getContentUri(CONTENT_URI_NAME, id);
+        String[] projection = new String[]{MediaStore.Audio.Playlists.Members.PLAY_ORDER};
+        Cursor cursor = mContentResolver.query(uri, projection, null, null, null);
+        int base = FIRST_ELEMENT_INDEX;
+        if (cursor.moveToLast()) {
+            base = cursor.getInt(FIRST_ELEMENT_INDEX);
+        }
+        cursor.close();
+        return base;
+    }
+
+    @Override
+    public long createPlaylist(String name) {
+        long id = getPlaylistId(name);
+
+        if (id == ERROR_INDEX) {
```

Comment:
về cơ bản em nên tách method này ra thành
isExistPlaylist(int playlistID)  {
    return playlistID >=0;
}
```suggestion
        if (!isExistPlaylist(id) ) {
```
và ở đây gọi ntn thì nó clear hơn.